docs-infra: ngmodule pages

petebacondarwin opened the pull request.

petebacondarwin committed:
build(docs-infra): remove unused info-bar API template

petebacondarwin committed:
build(docs-infra): separate NgModules from Classes in API docs

mary-poppins commented:
You can preview 9b00818 at https://pr25734-9b00818.ngbuilds.io/.

petebacondarwin commented:
Some examples:

* https://pr25734-9b00818.ngbuilds.io/api/common/CommonModule

![screen shot 2018-08-30 at 07 43 13](https://user-images.githubusercontent.com/15655/44834237-736d1c80-ac28-11e8-8577-c81f14ead56e.png)


* https://pr25734-9b00818.ngbuilds.io/api/forms/FormsModule

![screen shot 2018-08-30 at 07 43 33](https://user-images.githubusercontent.com/15655/44834243-77993a00-ac28-11e8-9c86-e78a54309c66.png)

IgorMinar commented on aio/tools/transforms/templates/api/ngmodule.template.html:
```diff
@@ -0,0 +1,66 @@
+  {$ renderTable(doc.ngmoduleOptions.providers, 'providers', 'Providers', 'Provider') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.imports %}
```
imports is an implementation detail of the module and should not be part of the public api docs. please remove.

IgorMinar commented on aio/tools/transforms/templates/api/ngmodule.template.html:
```diff
@@ -0,0 +1,66 @@
+  {$ renderTable(doc.ngmoduleOptions.declarations, 'declarations', 'Declarations', 'Declaration') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.entryComponents %}
```
private stuff, please remove

IgorMinar commented on aio/tools/transforms/templates/api/ngmodule.template.html:
```diff
@@ -0,0 +1,66 @@
+  {$ renderTable(doc.ngmoduleOptions.exports, 'exports', 'Exports', 'Export') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.declarations %}
```
private implementation detail, please remove

IgorMinar reviewed (commented):
I've thought a bit about "exports" because that is what represent the public api surface of the contents of an NgModule in respect to the user of the module.

The challenge is that in most cases, we export stuff indirectly via arrays and providers and many of these are private and shouldn't be listed in the public documentation. In the ideal case, we'd want to document that a module exports directives "A", "B" and "C", provides a provider for injection tokens "X", "Y", and "Z", and reexports another public NgModule, but this information is not trivial to extract without employing the logic from `ngc` and even then, we have internal NgModules that are not surfaced to the public api so even ngc doesn't have all the info to make this list.

I suggest that we don't even try to do that and simply just manually document these exports as three lists of symbols (directives, providers, reexported public NgModules) that then gets linked to the respective API pages and it will be up to humans to keep this list up to date. We don't changes this stuff very often so it should be ok to maintain these lists by hand.

petebacondarwin commented:
So @IgorMinar are you saying that I should also remove the `Exports` and `Providers` sections, and leave the author to manually describe these in the description/usage notes?

petebacondarwin committed:
fixup! build(docs-infra): separate NgModules from Classes in API docs

mary-poppins commented:
You can preview cf570d3 at https://pr25734-cf570d3.ngbuilds.io/.

mary-poppins commented:
You can preview 317f690 at https://pr25734-317f690.ngbuilds.io/.

jenniferfell commented:
One minor thing:  Can we choose another color for "NGMODULE" badge? We'd like to use red for re-instating the "deprecated" badge and make that more visible red (warning/caution) instead of its v5 grey.

jenniferfell commented:
notes from discussion 9/6...directives: new tagging like forms library so we can get to the directives and display them. "exported from this ngmodule". no tag = doesn't appear = internal directive. tag directives and pipes now. @ngmodule on every directive or pipe, if missing then error, linked ngmodule must be public or error, then surface on directives and pipes API pages. land this PR, and then new PR for the infra for tagging, followed by PR for actual tags added.

jenniferfell commented:
notes from discussion 9/6: providers: need more discussion. separate work.

petebacondarwin commented:
> One minor thing: Can we choose another color for "NGMODULE" badge? We'd like to use red for re-instating the "deprecated" badge and make that more visible red (warning/caution) instead of its v5 grey.

Happy to change the colour. Please choose one and let me know :-)

petebacondarwin committed:
docs: add missing `@ngModule` tags

petebacondarwin commented:
I have added two commits. The first adds the directive/pipe export rendering that we discussed. The second fixes up the last of the directives/pipes that were missing `@ngModule` tags - it turns out most already had them!

mary-poppins commented:
You can preview 6d879c9 at https://pr25734-6d879c9.ngbuilds.io/.

petebacondarwin commented:
Package pages now list NgModules too: https://pr25734-988458b.ngbuilds.io/api/forms

<img width="887" alt="screen shot 2018-09-08 at 21 48 44" src="https://user-images.githubusercontent.com/15655/45258518-063c5280-b3b1-11e8-9cc8-744666d09097.png">

petebacondarwin commented:
NgModules pages now list directives and pipes: https://pr25734-988458b.ngbuilds.io/api/common/CommonModule

<img width="892" alt="screen shot 2018-09-08 at 21 51 31" src="https://user-images.githubusercontent.com/15655/45258541-6b904380-b3b1-11e8-9436-ba2616cef6dd.png">

<img width="885" alt="screen shot 2018-09-08 at 21 51 40" src="https://user-images.githubusercontent.com/15655/45258542-71862480-b3b1-11e8-9386-4eb0673ef82a.png">

petebacondarwin commented:
Directives/Pipes pages list what NgModules they are exported from: https://pr25734-988458b.ngbuilds.io/api/common/NgClass

<img width="885" alt="screen shot 2018-09-08 at 21 54 20" src="https://user-images.githubusercontent.com/15655/45258574-c88bf980-b3b1-11e8-98f4-032f2c2fbad2.png">

In the case of shared directives that are exported from an internal NgModule that is then re-exported to more than one public NgModule, they appear as exports on both NgModule pages and list both modules on their own page: https://pr25734-988458b.ngbuilds.io/api/forms/RequiredValidator

<img width="882" alt="screen shot 2018-09-08 at 21 55 39" src="https://user-images.githubusercontent.com/15655/45258586-00933c80-b3b2-11e8-855f-d2b50e929b50.png">

petebacondarwin committed:
build(docs-infra): process and render ngmodule exports

All directives and pipes must now be tagged with one ore more
public NgModule, from which they are exported.

If an item is exported transitively via a re-exported internal NgModule
then it may be that the item appears to be exported from more than
one public NgModule. For example, there are shared directives that
are exported in this way from `FormsModule` and `ReactiveFormsModule`.

The doc-gen will error and fail if a directive or pipe is not tagged correctly.

NgModule pages now list all the directives and pipes that are exported from it.
Directive and Pipe pages now list any NgModule from which they are exported.
Packages also now list any NgModules that are contained - previously they were
missed.

mary-poppins commented:
You can preview f25f83a at https://pr25734-f25f83a.ngbuilds.io/.

IgorMinar commented:
All these changes look good and are way better than what I originally proposed. We could do a bit more of visual tweaking but it's fine if that's a follow up PR.

In case of public NgModules that are reexported via another public NgModule, it would be nice to list the reexported module(s). For example https://pr25734-f25f83a.ngbuilds.io/api/platform-browser/BrowserModule reexports CommonModule.

I suggest that we get this PR green and land it and then iterate on top of it. Thanks Pete!

petebacondarwin commented:
@IgorMinar can you deal with the G3 presubmit requirement?

petebacondarwin commented:
Also it needs your approval.
